### DISCUSSION TITLE
Article review: rtx + direnv + python

### DISCUSSION THREAD

**Discussion by Crocmagnon** [jdx/mise#235] [Show and tell]
Hello there!

I wrote a draft article for my blog which I'd like to share with you in the hope that somebody could review it before I hit publish. It's fairly short (2-3 min max), I just want to avoid publishing anything obviously wrong.

Here's the link: https://gabnotes.org/setup-rtx-with-direnv-for-python/?draft_key=ac6f7ada-cec3-4aa0-90b2-21207c951a2b

I'd be thankful to get feedback from the maintainers, I intend to publish it this week if possible.

Also, the license for the post is very permissive so if by any chance you want to reuse anything, please do!

Finally: thanks for this cool piece of software! 😁 

Thanks a lot 🙏🏻

**Comment by Crocmagnon** [comment]
Oh I just noticed my setup doesn't work as intended >.<

The virtualenv isn't properly activated, rtx's python version takes precedence in the PATH.

**Reply by Crocmagnon** [reply]
Fixed it in the article

**Reply by Crocmagnon** [reply]
I'm not veeeery thrilled by this setup though. I must delegate all rtx activating to direnv, even when not in a project directory. Is there a better way?

**Reply by jdx** [reply]
I wouldn't recommend using `rtx direnv activate`, that's more of a compatibility mode or for people that don't want rtx modifying PATH directly. It has less functionality and requires an .envrc file next to any .tool-versions file.

I would just use rtx like normal with `rtx activate`. As long as it's referenced in the rc file before `direnv hook` everything should work fine.

**Reply by Crocmagnon** [reply]
Thanks ! I was under the impression that the readme clearly encouraged setting up rtx’s hook *after* direnv’s. Did I misread something?

I’ll test that when I get back at my computer.

**Reply by jdx** [reply]
I was wrong just now, rtx needs to run before direnv but to do that you counterintuitively put the rtx hook after the direnv one. Technically what I said was correct (rtx runs after direnv), but the way you configure that is by reversing them.

**Reply by jdx** [reply]
I actually did some testing and things weren't working the way I had expected. I've changed the recommendation to use `eval "$(rtx env -s bash)"` if you need to use `layout python`.

**Reply by Crocmagnon** [reply]
Ok so I updated rtx to `1.19.0 macos-arm64 (built 2023-02-28)`, I have direnv and rtx setup in my fish config and rtx env in my envrc:

```
# config.fish
# ...
direnv hook fish | source
rtx activate fish | source
```

```
# project-local .envrc
eval "$(rtx env -s bash)"
layout python3
```

I removed `use rtx` from `~/.config/direnv/direnvrc`

now the virtualenv version doesn't get properly activated, i.e.
```
❯ which python
/Users/gaugendre/.local/share/rtx/installs/python/3.11.2/bin/python
```

instead of direnv's venv version.

`rtx env -s bash` outputs the global install first and direnv's venv just after.

```
❯ rtx env -s bash
export PATH='/Users/gaugendre/.local/share/rtx/installs/python/3.11.2/bin:/Users/gaugendre/Projects/charasheet/.direnv/python-3.11.2/bin:/Users/gaugendre/.local/share/rtx/installs/python/3.11.2/bin:/Users/gaugendre/.krew/bin:/Users/gaugendre/.cargo/bin:/Users/gaugendre/.local/bin:/opt/homebrew/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin:/usr/local/go/bin:/Library/TeX/texbin'
```

Well it turns out it's twice in the path in fact. Here's the same command run from outside the project's dir:
```
❯ rtx env -s bash
export PATH='/Users/gaugendre/.local/share/rtx/installs/python/3.10.7/bin:/Users/gaugendre/.krew/bin:/Users/gaugendre/.cargo/bin:/Users/gaugendre/.local/bin:/opt/homebrew/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin:/usr/local/go/bin:/Library/TeX/texbin'
```

The only setup currently working for me is the one described at the moment in the article: no `rtx activate`, `use rtx` in `direnvrc` and `layout python3` in `.envrc`.

**Reply by jdx** [reply]
I think my change in https://github.com/jdxcode/rtx/pull/204 to fix tmux broke this, let me test and see

**Reply by jdx** [reply]
does it matter if the python it points to isn't the one in the venv if they're the same version? If so, I don't think `rtx activate` will work for `layout python`. There will always be situations where rtx will update PATH to use the rtx version.

**Reply by Crocmagnon** [reply]
Yes it does. The Python interpreter must be the one in the venv.

**Reply by jdx** [reply]
ok then I should update the docs since you won't want to use `rtx activate` with `layout python` since it won't guarantee that direnv's config will override rtx.

**Reply by Crocmagnon** [reply]
Ok so the setup in the article is valid then? No `rtx activate` and `use rtx` in `.envrc` or `direnvrc`?

**Reply by jdx** [reply]
yeah that should work well. I suppose if you wanted to use the experimental shims support you could get rid of the restriction that every .tool-versions has to be paired with an .envrc file, and shims in rtx are a lot faster than they are in pyenv, but I realize that's exactly what you want to avoid with rtx.

Hopefully one day I can sort this out and it'll work more seamlessly, but for now I think you probably don't want to use `layout python` with `rtx activate`.

**Reply by Crocmagnon** [reply]
Yeah, I was never a fan of shims, always found them confusing. I guess I prefer good old PATH manipulation 😁 

Thanks a lot for your help! 👍🏻 😃

**Reply by jdx** [reply]
> Yeah, I was never a fan of shims, always found them confusing. I guess I prefer good old PATH manipulation 😁

you and me both

**Reply by amoosbr** [reply]
For my everyday use, I also switched to `direnv hook` and `use rtx` managed PATH handling (in `~/.config/direnv/direnvrc`).

The indeterministic behavior of `rtx activate` in combination `direnv hook` regarding PATH ordering was the main reason. And as mentioned in the original issue, as long as the rtx hook cannot identify which non-rtx PATH segments should stay before rtx-managed PATH segments and which after, this probably will continue to be an issue.
Calling `touch .envrc && direnv allow` for each `.tool-versions` is for me less of an inconvenience.

Switching to new rtx config format and tool plugins is not yet an option in my environment, as many colleagues are also using asdf or other tools that should continue to work.

I use the rtx shims support for the MacOS GUI apps (IDEs), but also wanted to avoid them in my shell.

Initially, I thought a global `~/.envrc` file would be too slow on every terminal start (bad experience from the asdf-direnv times), but it's actually not an issue with rtx anymore.

The other inconvenience for me was, that direnv always printed the config on terminal load. As I didn't want to completely disable direnv logging on project level, I was able to work-around this by adding the following to my `.zshrc` file:
```sh
# set ASDF_DATA_DIR to rtx data dir to get p10k asdf option working with rtx
# https://github.com/romkatv/powerlevel10k/blob/master/internal/p10k.zsh
# also requires asdf to be on PATH. so just symlink to rtx `ln -s /opt/homebrew/bin/rtx asdf` even though symlinking asdf to rtx is not supported
export ASDF_DATA_DIR="$HOME/.local/share/rtx"

# rtx in combination with direnv calling `use rtx` in ~/.config/direnv/direnvrc
# call rtx hook-env to put JAVA_HOME, GOROOT already on environment before direnv starts
# just a convenience, to see less direnv +JAVA_HOME, +GOROOT direnv logs
eval "$(rtx hook-env)"
if [[ $PWD = $HOME ]]; then
  export DIRENV_LOG_FORMAT=
  emulate zsh -c "$(direnv export zsh)"
  unset DIRENV_LOG_FORMAT
else
  # print direnv environment, when new login shell is opened in non $HOME directory
  emulate zsh -c "$(direnv export zsh)"
fi
```

And to trigger a direnv reload in projects, when the global `.tool-versions` changed, I needed to add `watch_file /Users/user/.tool-versions` to `~/.config/direnv/direnvrc`

With this setup, I'm, for the moment, quite happy.

**Comment by BuildBackBuehler** [comment]
Hey I like the article, I'd say if anything it could go into more detail. 

As a noob, right now I'm scratching my head wondering how exactly everything plugs and plays together. 

Intend to use nodejs/pnpm/python/poetry

I'm unsure if I should add or install these globally. If I do so globally, it (Poetry specifically) seems to do what you explicitly looked to avoid – being in a virtualenv everywhere. 

I think it may just be a convolution but is this an issue because I used "rtx install python/poetry -g"? I'm thinking the -g causes them to be quote-en-quote "activated" and thus creating the virenv issue. Imagining "rtx install poetry" just does exactly that – install it systemically, in rtx, to be readily "activated" wherever I need it to be in a more local manner.
